### PR TITLE
Drop in Unidata xarray workshop notebook

### PR DESCRIPTION
**Pull request by dcamron**
This is a drop-in of [this notebook](https://github.com/Unidata/python-training/blob/master/pages/workshop/XArray/XArray%20and%20CF.ipynb) from our workshop materials with interactive sections replaced with equivalent demonstrations.

Some notes:
 - Notebook definitely has crossover with data formats and map projections out of the box. The scope can be adjusted, but making relevant content that ties xarray to e.g. NetCDF will of course be important to us anyway.
 - Requires committing some example data, no remote-access content currently (also important)
 - Needs some reorganization and more jupyter-book friendly title and image formatting

### CODE REVIEW

File: core/xarray/xarray-nc-cf.ipynb [hunk 1]
```diff
@@ -0,0 +1,864 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# XArray & CF Introduction\n",
+    "\n",
+    "<div style=\"float:right; width:250 px\"><img src=\"http://xarray.pydata.org/en/stable/_static/dataset-diagram-logo.png\" alt=\"xarray Logo\" style=\"height: 250px;\"></div>\n",
+    "\n",
+    "### Questions\n",
+    "1. What is XArray?\n",
```
Comment (jukent) [inline core/xarray/xarray-nc-cf.ipynb:12]: On the xarray documentation, when capitalized it seems to always be "Xarray" not "XArray." Is there a reason for this choice? We'll want to be consistent across the site.

Comment (dopplershift) [reply core/xarray/xarray-nc-cf.ipynb:12]: I'm sure I got it wrong when I wrote the original notebook. (At least I think it was me.) I agree, based on the Xarray documentation, the canonical spelling seems to be "Xarray".

File: core/xarray/xarray-nc-cf.ipynb [hunk 2]
```diff
@@ -0,0 +1,864 @@
+{
```
Comment (ktyle) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>(just my initial test of ReviewNB) How about including making "Common Data Model (CDM)" a clickable link, pointing to https://www.unidata.ucar.edu/software/netcdf-java/v4.3/CDM/ ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Comment by dcamron** [comment]
I've pinged @ProjectPythia/education for review (I'll try to publicly self-review the content too.) As discussed in #42 and today's meeting notes, we've decided that **this PR** should be our main hub for focused review of a core content notebook. By our next meeting 5/13, we can plan to have this be our best current take at what we want our demonstrative notebook material to look like in terms of pacing, scope, presentation, and organization. We can then take this notebook and turn it into the template we generally plan to apply to the rest of our core teaching notebooks, as well as to provide for outside contributors to this book.

**Comment by brian-rose** [comment]
Just fixed a conflict in `environment.yml`

File: core/xarray/xarray-nc-cf.ipynb (hunk 2 above)
Comment (clyne) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>A minor nit mostly to check out ReviewNB: How about doing an actual assignment here:</p><p> </p><p>temp_in_celsius = temp - 273.15</p><p> </p><p>or explicitly pointing out that the values were not saved.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (clyne) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>This statement is formatted literally as ".interp(docs)". Is that intentional?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (clyne) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>Maybe display the value of "times" above and/or explain more what the calculation of times is doing. The code is pretty dense. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (clyne) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>"need to first need" -&gt; "we first need to"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (clyne) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>Why go ahead and make the 1D coordinate variables CF "Coordinate Variables", i.e. name them after the dimensions? It seems like that would simply things in the later discussion (and encourage good habits?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

Comment (clyne) [inline core/xarray/xarray-nc-cf.ipynb:1]: <p>I don't see where in the code you are setting the Temperature:coordinates variable as described above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='74'/>

**Comment by clyne** [comment]
Most of my in-line comments are small nits. One more substantive comment is that reading and writing CF compliant NetCDF files is an important and complex topic. I'm wondering if a more concise and simple example should be used here to simply demonstrate read/writing data from NetCDF (without CF compliance), and having a separate book (chapter?) that gives more in depth coverage of this topic. Several examples could be included starting from the most simple, but useful CF compliant file (e.g. 2D, no projection, static) and working up to the more complex (e.g. 3D, map projection, time-varying). I think this area is one that is not well-covered elsewhere and is vitally important for folks working with CF data. What do others think?

**Comment by ktyle** [comment]
With the expectation that today's meeting will focus on this excellent notebook, here is a "core dump" of my thoughts ...

1. The overall flow and pedagogical nature of the notebook is great!
2. I suggest splitting the notebook as it runs a bit long, particularly for one that is intended as an introduction to the library. A natural breakpoint would be when the section on the CF conventions begins.
3. From the get-go, I would include a discussion of a Dataset in addition to the DataArray. In my mind at least, the Dataset and DataArray are the two "workhorses" of Xarray ... just as a Series and DataFrame are for Pandas.
4. Does it make sense to include the NARR dataset for the examples (I actually do think this is a good idea, btw)? Any reason why we wouldn't defer to the air temperature dataset that is used in the Xarray package documentation examples? 
5. The initial notebook should also include at least a couple of visualizations, since, like Pandas, Xarray has a built-in interface to Matplotlib.
6. I suspect that as we populate sections on core scientific python packages that come earlier in the sequence, such as Matplotlib and Pandas, sections that come later (such as Xarray) will need to be updated so we can add content that reiterates and reinforces material that come in the earlier sections. The CF convention material, for example, should nicely benefit from what we place in the "Data Formats" section. 
7. I am going to slightly rework some of the intro to Xarray notebooks that I have used in the two classes I teach and then submit them to the group for review. I would imagine that some of the material that I have developed will nicely complement this particular notebook.

**Comment by brian-rose** [comment]
I agree with much of @ktyle's comments here. For pedagogical purposes, I support moving the CF conventions to a different notebook, and splitting the xarray content up with some more headings and shorter sections.

Two key features of xarray (to my mind) that don't appear in this notebook but could are:
1. automatically labeled plots
2. arithmetic between two `DataArray` objects to demonstrate how xarray uses coordinates to handle broadcasting.

To me point 2 is the real "killer feature" that shows the huge advantage of using xarray over plain numpy arrays when working with gridded data. I think it would be great to slip this in somewhere early on in our tutorials.

**Comment by mgrover1** [comment]
Retweet what @ktyle and @brian-rose mentioned - I think maybe mentioning cf-xarray within the cf notebook would be good too.

Maybe a quick example of an arithmetic example, detailing how it deals with dimensionality of the datasets.

**Comment by jukent** [comment]
There are still lots of comments on this PR (Xarray is still XArray throughout the document, for one). I will do a thorough review after these conversations are resolved.

**Comment by clyne** [comment]
@brian-rose will take ownership

**Comment by brian-rose** [comment]
A link to scipy.org is failing.

File: core/xarray/xarray-nc-cf.ipynb (hunk 2 above)
Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>That link appears to be dead, but I linked to https://docs.unidata.ucar.edu/netcdf-java/current/userguide/common_data_model_overview.html</p>

Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>Revised to do an actual assignment here, with a little more verbose discussion.</p>

Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>I cleaned this up.</p>

Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>Removing the CF material from this notebook.</p>

Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>Removing the CF material from this notebook.</p>

Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>Removing the CF material from this notebook.</p>

Comment (brian-rose) [reply core/xarray/xarray-nc-cf.ipynb:1]: <p>Removing the CF material from this notebook.</p>

**Comment by brian-rose** [comment]
Ok team, I have substantially revised this Xarray notebook based on all reviews:
- Adapted to our template format
- Removed the CF conventions material (actually I just copied it into a "hidden" notebook that's in the repo but not rendered in the book, to be dealt with later
- Added an earlier introduction to the `Dataset` object
- Added some visualization examples

It's not perfect, but hopefully good enough to get something merged in the next 12 hours or so.

I'm happy to do small corrections tomorrow morning. If you have more existential issues with this notebook, it may be best to open new issues for later work.

**Comment by clyne** [comment]
Wiill review asap this a.m.

**Comment by brian-rose** [comment]
Assuming this gets merged in form close to what's currently here, we should open a new issue about sorting out the CF material, which is currently in a hidden notebook on this branch at `core/xarray/nc-cf.ipynb`

File: core/xarray/xarray-intro.ipynb [hunk 3]
```diff
@@ -0,0 +1,935 @@
...
+    "from pythia_datasets import DATASETS  # Project Pythia's custom store of example data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introducing the `DataArray` and `Dataset`\n",
+    "\n",
+    "Xarray expands on the capabilities on NumPy arrays, providing a lot of streamlined data manipulation. It is similar in that respect to Pandas, but whereas Pandas excels at working with tabular data, XArray is focused on N-dimensional arrays of data (i.e. grids). Its interface is based largely on the netCDF data model (variables, attributes, and dimensions), but it goes beyond the traditional netCDF interfaces to provide functionality similar to netCDF-java's [Common Data Model (CDM)](https://docs.unidata.ucar.edu/netcdf-java/current/userguide/common_data_model_overview.html). "
```
Comment (jukent) [inline core/xarray/xarray-intro.ipynb:99]:
```suggestion
    "Xarray expands on the capabilities on NumPy arrays, providing a lot of streamlined data manipulation. It is similar in that respect to Pandas, but whereas Pandas excels at working with tabular data, Xarray is focused on N-dimensional arrays of data (i.e. grids). Its interface is based largely on the netCDF data model (variables, attributes, and dimensions), but it goes beyond the traditional netCDF interfaces to provide functionality similar to netCDF-java's [Common Data Model (CDM)](https://docs.unidata.ucar.edu/netcdf-java/current/userguide/common_data_model_overview.html). "
```

File: core/xarray/xarray-intro.ipynb [hunk 4]
```diff
@@ -0,0 +1,935 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Attributes are not preserved by default!\n",
+    "\n",
+    "Notice what happens if we perform a mathematical operaton with the `DataArray`: the coordinate values persist, but the attributes are lost. This is done because it is very challenging to know if the attribute metadata is still correct or appropriate after arbitrary arithmetic operations.\n",
+    "\n",
+    "To illustrate this, we'll do a simple unit conversion from Kelving to Celsius:"
```
Comment (jukent) [inline core/xarray/xarray-intro.ipynb:280]:
```suggestion
    "To illustrate this, we'll do a simple unit conversion from Kelvin to Celsius:"
```

File: core/xarray/xarray-intro.ipynb [hunk 5]
```diff
@@ -0,0 +1,935 @@
...
+    "## Resources and references\n",
+    "\n",
+    "This notebook was adapated from material in [Unidata's Python Training](https://unidata.github.io/python-training/workshop/XArray/xarray-and-cf/)\n",
+    "\n",
+    "The best resource for Xarray is the [Xarray documentation](http://xarray.pydata.org/en/stable/). See in particular\n",
+    "- [Why Xarray](http://xarray.pydata.org/en/stable/getting-started-guide/why-xarray.html)\n",
+    "- [Quick overview](http://xarray.pydata.org/en/stable/getting-started-guide/quick-overview.html#)\n",
+    "- [Example gallery](http://xarray.pydata.org/en/stable/gallery.html)\n",
+    "\n",
+    "Another excellent resource is this [Xarray Tutorial collection](https://xarray-contrib.github.io/xarray-tutorial/)"
```
Comment (jukent) [inline core/xarray/xarray-intro.ipynb:903]:
```suggestion
    "Another excellent resource is this [Xarray Tutorial collection](https://xarray-contrib.github.io/xarray-tutorial/)."
```

File: core/xarray/xarray-intro.ipynb [hunk 6]
```diff
@@ -0,0 +1,935 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Resources and references\n",
+    "\n",
+    "This notebook was adapated from material in [Unidata's Python Training](https://unidata.github.io/python-training/workshop/XArray/xarray-and-cf/)\n",
```
Comment (jukent) [inline core/xarray/xarray-intro.ipynb:896]:
```suggestion
    "This notebook was adapated from material in [Unidata's Python Training](https://unidata.github.io/python-training/workshop/XArray/xarray-and-cf/).\n",
```

**Review by jukent — CHANGES_REQUESTED** [review]
This is a big improvement! Just a few small changes.

File: core/xarray/xarray-intro.ipynb [hunk 7]
```diff
@@ -0,0 +1,935 @@
+{
```
Comment (clyne) [inline core/xarray/xarray-intro.ipynb:1]: <p>This example would be much more clear if you pointed out that the time sample 2018-01-07 is not present in the temp array</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='46'/>

Comment (clyne) [inline core/xarray/xarray-intro.ipynb:1]: <p>Ditto. Pointing out lon=-105 is not one of the sampled longitude coordinates.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (clyne) [inline core/xarray/xarray-intro.ipynb:1]: <p>Sum or std?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ProjectPythia/pythia-foundations/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='75'/>

**Review by clyne — COMMENTED** [review]
Much improved, Brian! I just had a few nits, then should be good to go.

**Comment by brian-rose** [comment]
Thanks for the reviews @jukent and @clyne. I'll get the suggested changes pushed up shortly.

File: core/xarray/xarray-intro.ipynb (hunk 4 above)
Comment (brian-rose) [reply core/xarray/xarray-intro.ipynb:280]: Done

File: core/xarray/xarray-intro.ipynb (hunk 5 above)
Comment (brian-rose) [reply core/xarray/xarray-intro.ipynb:903]: Done.

File: core/xarray/xarray-intro.ipynb (hunk 6 above)
Comment (brian-rose) [reply core/xarray/xarray-intro.ipynb:896]: Done.

File: core/xarray/xarray-intro.ipynb (hunk 7 above)
Comment (brian-rose) [reply core/xarray/xarray-intro.ipynb:1]: <p>Good, I added some extra explanation about this.</p>

Comment (brian-rose) [reply core/xarray/xarray-intro.ipynb:1]: <p>Great, I added some preamble to this example to explain what we're doing and why.</p>

Comment (brian-rose) [reply core/xarray/xarray-intro.ipynb:1]: <p>Good catch. It should be <code>std</code>.  I fixed this, and also added a new Note callout to draw attention to the use of the <code>dim=</code>  keyword (another important use of named coordinates).</p>

**Comment by brian-rose** [comment]
@clyne @jukent as soon as this builds, it's ready for a final review and approval.

**Review by jukent — APPROVED** [review]

**Comment by clyne** [comment]
In meetings until 11:30am MT, but will jump right on it then

**Review by clyne — APPROVED** [review]
Awesome! Approved and merging.

**Pull request merged**